src/GHC/TypeLits/Extra/Solver.hs

```diff
@@ -114,100 +115,127 @@ decideEqualSOP _    _givens _deriveds []      = return (TcPluginOk [] [])
       where
         go [] = Nothing
-        go ((ct, a,b@(Max _ _)) :_)
+        go ((NatEquality ct a b@(Max _ _)) :_)
           | c == a && not (isWantedCt ct)
             = Just b
-        go ((ct, a@(Max _ _),b) :_)
+        go ((NatEquality ct a@(Max _ _) b) :_)
           | c == b && not (isWantedCt ct)
             = Just a
         go (_:rest) = go rest
 
 
 -- Extract the Nat equality constraints
-toNatEquality :: ExtraDefs -> Ct -> MaybeT TcPluginM (Either NatEquality NatInEquality)
-toNatEquality defs ct = case classifyPredType $ ctEvPred $ ctEvidence ct of
+toSolverEquality :: ExtraDefs -> Ct -> MaybeT TcPluginM SolverConstraint
+toSolverEquality defs ct = case classifyPredType $ ctEvPred $ ctEvidence ct of
     EqPred NomEq t1 t2
       | isNatKind (typeKind t1) || isNatKind (typeKind t2)
-      -> Left <$> ((ct,,) <$> normaliseNat defs t1 <*> normaliseNat defs t2)
+      -> do
+         (t1', n1) <- normaliseNat defs t1
+         (t2', n2) <- normaliseNat defs t2
+         if n1 || n2
+           then lift $ NormalisedCt ct <$> createDerivedWithNormalised ct (reifyEOP defs t1') (reifyEOP defs t2')
+           else pure (NatEquality ct t1' t2')
       | TyConApp tc [x,y] <- t1
       , tc == typeNatLeqTyCon
       , TyConApp tc' [] <- t2
-      -> if tc' == promotedTrueDataCon
-            then Right <$> ((ct,,,True) <$> normaliseNat defs x <*> normaliseNat defs y)
-            else if tc' == promotedFalseDataCon
-                 then Right <$> ((ct,,,False) <$> normaliseNat defs x <*> normaliseNat defs y)
-                 else fail "Nothing"
+      -> do
+          (x', n1) <- normaliseNat defs x
+          (y', n2) <- normaliseNat defs y
+          case (tc', n1 || n2) of
+            (_, True)                            -> let t1' = TyConApp tc [reifyEOP defs x', reifyEOP defs y']
+                                                    in lift $ NormalisedCt ct <$> createDerivedWithNormalised ct t1' t2
+            (_, _) | tc' == promotedTrueDataCon  -> pure (NatInEquality ct x' y' True)
+            (_, _) | tc' == promotedFalseDataCon -> pure (NatInEquality ct x' y' False)
+            _                                    -> fail "Nothing"
     _ -> fail "Nothing"
   where
     isNatKind :: Kind -> Bool
     isNatKind = (`eqType` typeNatKind)
 
-fromNatEquality :: Either NatEquality NatInEquality -> Ct
-fromNatEquality (Left (ct, _, _))  = ct
-fromNatEquality (Right (ct,_,_,_)) = ct
+createDerivedWithNormalised :: Ct -> Type -> Type -> TcPluginM CtEvidence
+createDerivedWithNormalised ct t1 t2 = do
+  newPredTy <- case splitTyConApp_maybe $ ctEvPred $ ctEvidence ct of
+    Just (tc, [a, b, _, _]) | tc `hasKey` eqPrimTyConKey -> pure (mkTyConApp tc [a, b, t1, t2])
+    _ -> fail "Nothing"
+  newWanted (ctLoc ct) newPredTy
+
+fromSolverEquality :: SolverConstraint -> Ct
+fromSolverEquality (NatEquality ct _ _)  = ct
+fromSolverEquality (NatInEquality ct _ _ _) = ct
+fromSolverEquality (NormalisedCt ct ev) = ct { cc_ev = ev }
```

Comment:
This is kind of dirty? I use the old constraint and only the change the `CtEvidence` field.